### ISSUE TITLE
Move role creation into the recipe instead of feature activation

### ISSUE THREAD

**Issue by MikeAlhayek** [OrchardCMS/OrchardCore#12425]
### Is your feature request related to a problem? Please describe.
Currently, we create roles based on the registered Permission Providers when a feature is installed. This approach has the following two disadvantages 
 1. There is no way to control which roles to add when the site is setup. Some admins don't want to `Contributor`, `Moderator` or any other role. In today's setup, they can't opt-out of these roles even if they don't want them.
 2. If the site admin delete a role using the UI, this role could come back up when he/she installs the next feature. If an admin delete a role, I don't think we should add it back up just because we enabled a new feature. This also create confusion (like why this role is back?!)

IMO, if an admin choose to opt-out of having a role added to their website because they are handling permission grouping differently, then we should not be forcing them into using some default permission.

### Describe the solution you'd like
Instead of adding roles during feature setup, add the role should be part of the setup recipes. However, we keep today's logic for the permissions. So when the site is first setup, we first execute the [RolesStep](https://github.com/OrchardCMS/OrchardCore/blob/339d74b251b53393dd510f8ad3027aa17d78500d/src/OrchardCore.Modules/OrchardCore.Roles/Recipes/RolesStep.cs) to load the needed role names. Then in the [RoleUpdater](https://github.com/OrchardCMS/OrchardCore/blob/339d74b251b53393dd510f8ad3027aa17d78500d/src/OrchardCore.Modules/OrchardCore.Roles/Services/RoleUpdater.cs#L83-L84), we skip creating the role when the roles does not exists.

We may need to add a new `int Order` property to `IRecipeStepHandler` to ensure that `RolesStep` in migrated before [FeatureStep](https://github.com/OrchardCMS/OrchardCore/blob/4192c145724186c442715414f16d98e883c4b0b1/src/OrchardCore.Modules/OrchardCore.Features/Recipes/Executors/FeatureStep.cs) since [FeatureStep](https://github.com/OrchardCMS/OrchardCore/blob/4192c145724186c442715414f16d98e883c4b0b1/src/OrchardCore.Modules/OrchardCore.Features/Recipes/Executors/FeatureStep.cs) triggers that events used by [RoleUpdater](https://github.com/OrchardCMS/OrchardCore/blob/339d74b251b53393dd510f8ad3027aa17d78500d/src/OrchardCore.Modules/OrchardCore.Roles/Services/RoleUpdater.cs#L83-L84).

**Comment by MikeAlhayek** [comment]
@jtkech the only problem I am having with the above approach is that the [RoleUpdater](https://github.com/OrchardCMS/OrchardCore/blob/339d74b251b53393dd510f8ad3027aa17d78500d/src/OrchardCore.Modules/OrchardCore.Roles/Services/RoleUpdater.cs#L83-L84), gets called during tenant setup which for some reason happens before recipes are executed. For this to work, importing roles must happen before `IFeatureEventHandler.InstalledAsync(IFeatureInfo feature)` events are called. 

But I am not sure how `IFeatureEventHandler.InstalledAsync(IFeatureInfo feature)` is called before recipes are executed? The `FeatureStep` should have be also executed before the `IFeatureEventHandler.InstalledAsync` are called. Thoughts?

**Comment by jtkech** [comment]
From memory, each recipe step is executed in its own shellScope. When the features step has been executed, at the end of this step scope and before the following steps, a deferred task (in its own scope) calls the features handlers. The deferred task is triggered in `ShellDescriptorFeaturesManager` line 119.

So if we use the features step first, the feature handlers will be caled after this step, but before any other steps. That said normally `RoleUpdater` is only called if registered, so only if `OC.Roles` is enabled.

I say this because at some point you wanted to be able to not use `OC.Roles`, but maybe here you still want more control on roles even if `OC.Roles` is enabled.

Normally when enabling a feature the `RoleUpdater` only add roles tied to this feature (and its dependencies). Hmm but yes many features have an `IPermissionProvider` using e.g. the same `PermissionStereotype` as `Moderator`.

Hmm yes would need to have configurable role names, not necessarily through a recipe, maybe a default `IRoleNamesProvider` easily overridable, or from the configuration or only an IOptions class having a default set of roles that e.g. a `services.Configure<>()` could easily update, need more thinkings, maybe currently you just need to override/replace the `RoleUpdater`.

**Comment by MikeAlhayek** [comment]
@jtkech thanks for the info. Yes the role updater should be kept with roles feature and the site should not depend on roles.

Using configuration would solve half of the puzzle but would make things a bit easier.  I would probably store that info into the site settings just so we can allow different tenants to be constructed with a different roles. The remaining issue is what if an admin disabled roles for a while and then decided to enable it again: any feature was installed while roles was disabled, won't be assigned to any role. Also, what if and admin removed a specific permission from a specific role, on the next feature install that assignment will come back which isn't good.

I think we need a document to store all evaluated permission (PermissionHistory). So in the role updated when a permission exists in the PermissionHistory we won't evaluate it any more. Also the logic in the RoleUpdater would need to be executed on every module enabled not installed so that way we alway add any new permissions and assign them accordingly. Do you agree?

**Comment by MikeAlhayek** [comment]
@jtkech I updated my previous comment with little more details.

**Comment by jtkech** [comment]
Yes, many things to consider, would need more time ;)

Hmm I don't think using site settings is good e.g. in a scenario which doesn't use site settings but roles.

> if an admin disabled roles feature for a while and then decides to enable it again? Any feature was installed while roles was disabled, won't be assigned to any role

Yes, I understand

> Also, what if and admin removed a specific permission from a specific role, on the next feature install that assignment will come back which isn't good.

Normally a feature is installed once, only when the feature is first enabled. Hmm maybe you are already thinking about the below point where permissions would be assigned to roles on feature enabling.

> Also the logic in the RoleUpdater would need to be executed on every module enabled not installed so that way we always add any new permissions and assign them accordingly.

Yes, I understand but seems to be contradictory with the previous point.

Hmm maybe the concept of Active Roles, let roles assigned and persisted as is but filter them, not everywhere but where we don't want to expose them when calling methods like `GetRolesAsync()`. 

Hmm maybe having Excluded Roles, a given role being valid if not excluded.

**Comment by MikeAlhayek** [comment]
@jtkech thanks for your feedback again.

We'll what if a a module introduced new permissions after a module is installed? Will these permissions get added to a role currently? It does not sound that we account for that today.

Isn't there an event that we can hook into that is called after the tenant is initialized and bootstrapped "Like tenant Ready"? If so, this is probably a good place to hook into instead of the setup event. So any time the app is started, we look for new permissions to associate them with roles as long as they were not previously evaluated using the PermissionHistory document. 

We would still hook into the feature-installed event to also evaluate permission. But, in this case we would evaluate all permission from all features "not just the one we just installed" as long as the permission name isn't listed in the PermissionHistory. This way we guarantee that permission are always up today by tracking them in the PermissionHistory.

**Comment by jtkech** [comment]
> We'll what if a a module introduced new permissions after a module is installed? Will these permissions get added to a role currently? It does not sound that we account for that today.

Hmm yes you are right, there is no Permissions migrations, for now you need to tie these new permissions to new feature(s), can be the same module but new feature(s).

> Isn't there an event that we can hook into that is called after the tenant is initialized and bootstrapped

There is `ModularTenantEvents`, e.g. `ActivatedAsync()` which is called each time a tenant/shell is built/rebuilt, so on the first request on this tenant after the app startup or after each time we release the tenant, e.g. after saving some site setting, enabling/disabling a feature. Note: So it may be called multiple times during a setup recipe.

> So any time the app is started, we look for new permissions to associate them with roles as long as they were not previously evaluated using the PermissionHistory document.

Hmm, looks like the `RoleUpdater` already checks current and new permissions, but in your case would need to be called not only when installing a given feature.

Hmm, about PermissionHistory maybe we already have it, we already have a distributed cached `RolesDocument` containing all Roles and then their permissions / claims.

Would need to be more focused on all of this, I will if I have time.

For info just saw that `RoleManager` already uses `IRoleValidator<TRole>` that are called before persistence.

https://github.com/dotnet/aspnetcore/blob/136a6ae2768b4b02625b3bc55d7aa9289dcb282f/src/Identity/Extensions.Core/src/RoleManager.cs#L37-L41

**Comment by MikeAlhayek** [comment]
@jtkech no problem take your time. I appreciate all the feedback. The RoleDocument provides us the mapping of current role permission, it does not tell us if a permission was assigned to a role but then it was removed using the UI. So it's important that we don't reassign a permission to a role if an admin explicitly remove it from a specific role. The RolesDocument would still need to tell which are the active claims to grant the user when they sign in. The PermissionHistory would just tell us if we can assign the permission that is provided by IPermissionProvoder to the role as defined by the provider. Once a permission is in the history, we leave it up to the site admin to manage permission after the initial assignment.

I'll try hooking to the tenant event tomorrow and provide an update.

**Comment by MikeAlhayek** [comment]
PR is ready for review
